packages/gatsby-theme-docs/src/hooks/use-active-section.js

```diff
@@ -1,32 +1,97 @@
 import React from 'react';
-import { designSystem } from '@commercetools-docs/ui-kit';
 import useScrollSpy from './use-scroll-spy';
 
-const calculateOffset = () => {
-  const pxNumber = designSystem.dimensions.heights.header.replace(
-    /([0-9]+)px$/,
-    '$1'
-  );
-  return parseInt(pxNumber, 10);
-};
-
 const getSectionElements = () =>
   document.querySelectorAll('section[class^="section-h"]');
 
-const offset = calculateOffset();
+/**
+ * The function tests if a specific block element is in withing the viewport of the browser.
+ * In particular we consider in view an element when:
+ * 1. The top of the element is lower than the top margin of the viewport and it is also lower thant the bottom margin of the viewport
+ * 2. The top of the element is higher than the top margin of the viewport and the bottom part of the element is lower of the bottom margin of the viewport
+ */
+const isInViewport = (element) => {
+  const rect = element.getBoundingClientRect();
+
+  return (
+    (rect.top >= 0 &&
+      rect.top <
+        (window.innerHeight || document.documentElement.clientHeight)) ||
+    (rect.top < 0 && rect.bottom >= 0)
+  );
+};
+
+const calclulateActiveSection = (
+  visibleElements,
+  locationHash,
+  fallbackElement
+) => {
+  if (!visibleElements) {
+    // should not happen as we always have a section in view, in case we highlight a fallbackElement
+    return fallbackElement;
+  }
+  if (locationHash) {
+    // if the url has the a hash, find the matching hash in the visible elements (if exists)
+    // otherwise return the first visible element
+    return (
+      visibleElements.find(
+        (element) => element.id === `section-${locationHash.slice(1)}`
+      ) || visibleElements[0]
+    );
+  }
+  return visibleElements[0];
+};
 
 const useActiveSelection = () => {
   const [activeSection, setActiveSection] = React.useState();
+
   const onScroll = React.useCallback(() => {
     const sectionElements = getSectionElements();
-    let nextActiveSection;
+    const locationHash = window.location.hash;
+    const visibleSections = [];
     sectionElements.forEach((section) => {
-      if (section.getBoundingClientRect().top - offset < 0) {
-        nextActiveSection = section;
+      if (isInViewport(section)) {
+        visibleSections.push(section);
       }
     });
-    setActiveSection(nextActiveSection);
+    setActiveSection(
+      calclulateActiveSection(visibleSections, locationHash, sectionElements[0])
+    );
+    // by resetting the url hash it prevents some incorrect behaviours when the hash
+    // is changed but the page cannot scroll (bottom of the page)
+    if (window.location.hash) {
```

Comment:
See the inline comment @nkuehn 
Not sure if you're going to like it.
This solves all the glitches with the section navigation menu but hides the hash from the url.
The hash functionality still works as before, it is just the URL that gets rewritten just after removing the hash. The UX result is that when pressing a link you don't see the hash in the url.
I don't think it's a biggie, we still have the copy url functionality in place, but wanted to check

Reply:
Your gut feeling is right, that's not how links in a content website should work IMHO, especially we can't have them behave inconsistently depending on whether a link leads to a new microsite, a new page or the same page.    

You'll need to explain me what the "glitches" would be.

Reply:
reverted.